### PR TITLE
Adjust the sharpness of TAA's Catmull-Rom filter based on local area luminance

### PR DESCRIPTION
**Pull request by invertednormal**
This helps prevent ringing artifacts in high contrast areas while in motion. Due to Catmull-Rom's use of negative weights, it's possible to have a very high value neighbor sample with a negative weight completely overwhelm the result leading to a negative output. This change reduces the sharpness of the catmull rom filter in high contrast areas so the negative weights contribute less to the final result.

Due to the changes to use float3 instead of float4 in SampleCatmullRom5Tap(), the performance of TAA actually improves by about 10%.

Before:
image
After:
image

### CODE REVIEW

**Review by mrieggeramzn — APPROVED** [review]

**Review by mrieggeramzn — COMMENTED** [review]
I remember seeing Microsoft comment years ago on how avoiding carrying around the .a can reduce vgpr usage and improve perf. Nice to see it in action.

File: Gems/Atom/Feature/Common/Assets/Shaders/PostProcessing/Taa.azsl [hunk 1]
```diff
@@ -99,11 +99,20 @@ float3 YCoCgToRgb(float3 yCoCg)
...
 // samples as their influence is negligible.
-float4 SampleCatmullRom5Tap(Texture2D<float4> texture, SamplerState linearSampler, float2 uv, float2 textureSize, float2 rcpTextureSize, float sharpness)
+float3 SampleCatmullRom5Tap(Texture2D<float4> texture, SamplerState linearSampler, float2 uv, float2 textureSize, float2 rcpTextureSize, float sharpness)
 {
     // Think of sample locations in the 4x4 neighborhood as having a top left coordinate of 0,0 and
     // a bottom right coordinate of 3,3.
 
+    // Use the local area green value as a measurement of luminance, and tweak the sharpness
+    // based on the difference between the max and min values. This will reduce the contribution
+    // of negative weights in high contrast areas and help prevent ringing artifacts.
```
Comment (Adi-Amazon) [inline Gems/Atom/Feature/Common/Assets/Shaders/PostProcessing/Taa.azsl:109]: super nit: start with two lines stating of the problem it comes to solve.

File: Gems/Atom/Feature/Common/Assets/Shaders/PostProcessing/Taa.azsl [hunk 2]
```diff
@@ -99,11 +99,20 @@ float3 YCoCgToRgb(float3 yCoCg)
...
+    float4 gatheredGreen = texture.GatherGreen(linearSampler, uv, int2(0, 0));
+    float minGreen = min(gatheredGreen.x, min(gatheredGreen.y, min(gatheredGreen.z, gatheredGreen.w)));
+    float maxGreen = max(gatheredGreen.x, max(gatheredGreen.y, max(gatheredGreen.z, gatheredGreen.w)));
+    float diff = maxGreen - minGreen;
+    sharpness /= diff + 1.0;
```
Comment (Adi-Amazon) [inline Gems/Atom/Feature/Common/Assets/Shaders/PostProcessing/Taa.azsl:114]: Are you sure this operands is doing the following:
sharpness = sharpness / (diff + 1.0) 
and not: 
sharpness = sharpness / diff + 1.0

For clarity and safety I would change to:
sharpness /= (diff + 1.0f);

**Review by Adi-Amazon — APPROVED** [review]
Good idea!
Please verify or add parenthesis (preferable) to the operand.

File: Gems/Atom/Feature/Common/Assets/Shaders/PostProcessing/Taa.azsl (hunk 1 above)
Comment (invertednormal) [reply Gems/Atom/Feature/Common/Assets/Shaders/PostProcessing/Taa.azsl:109]: Do you mean what this function is trying to solve? (there's a description above the function definition) or what this particular piece of code is solving (meaning you suggest re-wording this comment to state the problem first)?

File: Gems/Atom/Feature/Common/Assets/Shaders/PostProcessing/Taa.azsl (hunk 2 above)
Comment (invertednormal) [reply Gems/Atom/Feature/Common/Assets/Shaders/PostProcessing/Taa.azsl:114]: hlsl follows the same operator precedence as c++, and all the assignment operators are very low precedence so they happen last (only thing lower precedence is a comma). I can add parenthesis though if that helps with clarity.

File: Gems/Atom/Feature/Common/Assets/Shaders/PostProcessing/Taa.azsl (hunk 1 above)
Comment (invertednormal) [reply Gems/Atom/Feature/Common/Assets/Shaders/PostProcessing/Taa.azsl:112]: Updated the comment to be more clear.

**Pull request merged**